pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,491 @@
+          }
+
+          if (!outlierSegments.isEmpty()) {
+            pinotTaskConfigsForTable.addAll(
+                createPinotTaskConfigs(outlierSegments, offlineTableName, maxNumRecordsPerTask, mergeLevel,
+                    mergeConfigs, taskConfigs));
+          }
+        }
+      }
+
+      // Write updated watermark map to zookeeper
+      try {
+        _clusterInfoAccessor.setMergeRollupTaskMetadata(mergeRollupTaskMetadata, expectedVersion);
+      } catch (ZkException e) {
+        LOGGER.error(
+            "Version changed while updating merge/rollup task metadata for table: {}, skip scheduling. There are multiple task schedulers for the same table, need to investigate!",
+            offlineTableName);
+        continue;
+      }
+      pinotTaskConfigs.addAll(pinotTaskConfigsForTable);
+      LOGGER.info("Finished generating task configs for table: {} for task: {}", offlineTableName, taskType);
+    }
+    return pinotTaskConfigs;
+  }
+
+  /**
+   * Validate table config for merge/rollup task
+   */
+  private boolean validate(TableConfig tableConfig, String taskType) {
+    String offlineTableName = tableConfig.getTableName();
+    if (tableConfig.getTableType() != TableType.OFFLINE) {
+      LOGGER.warn("Skip generating task: {} for non-OFFLINE table: {}, REALTIME table is not supported yet", taskType,
+          offlineTableName);
+      return false;
+    }
+
+    if (REFRESH.equalsIgnoreCase(IngestionConfigUtils.getBatchSegmentIngestionType(tableConfig))) {
+      LOGGER.warn("Skip generating task: {} for non-APPEND table: {}, REFRESH table is not supported", taskType,
+          offlineTableName);
+      return false;
+    }
+
+    TableTaskConfig tableTaskConfig = tableConfig.getTaskConfig();
+    Preconditions.checkState(tableTaskConfig != null);
+    Map<String, String> taskConfigs = tableTaskConfig.getConfigsForTaskType(taskType);
+    Preconditions.checkState(taskConfigs != null, "Task config shouldn't be null for table: {}", offlineTableName);
+    return true;
+  }
+
+  /**
+   * Check if the segment is merged for give merge level
+   */
+  private boolean isMergedSegment(OfflineSegmentZKMetadata segmentZKMetadata, String mergeLevel) {
+    return segmentZKMetadata.getCustomMap() != null && mergeLevel.equalsIgnoreCase(
+        segmentZKMetadata.getCustomMap().get(MergeRollupTask.SEGMENT_ZK_METADATA_MERGE_LEVEL_KEY));
+  }
+
+  /**
+   * Check if the merge window end time is valid
+   */
+  private boolean isValidMergeWindowEndTime(long windowEndMs, long bufferMs, String lowerMergeLevel,
+      MergeRollupTaskMetadata mergeRollupTaskMetadata) {
+    // Check that execution window endTimeMs <= now - bufferTime
+    if (windowEndMs > System.currentTimeMillis() - bufferMs) {
+      return false;
+    }
+    // Check that execution window endTimeMs <= waterMark of the lower mergeLevel
+    return lowerMergeLevel == null || windowEndMs <= mergeRollupTaskMetadata.getWatermarkMap().get(lowerMergeLevel);
```

Comment:
`watermark` for the previous level? I think by current logic, cold start for the first merge level is the only case that the previous `watermark` is null. For this case, the `lowerMergeLevel` is `null`, so this will not cause NPE.

Reply:
With current logic, the only case when it can fail is when the first merge level does not have valid merge window end time, and the second merge level has a shorter buffer time than the first one (this is mis-configuration, and we should probably validate this case).
But again, the logic is quite complicated, so I suggest to go with the simple logic to only update watermark when new task is scheduled, then this null check is required

Reply:
^^, We need `null` check now as the previous level could be skipped

Reply:
I see, updated.